### PR TITLE
PyodideRuntime should be one of the runtimes

### PR DESCRIPTION
**Pull request by madhur-tandon**
- For the file -- `base.ts`, `pyenv.ts`, `pyscript.ts`:

All references to `PyodideInterface` have been replaced with `Runtime` which has the available methods of running code asynchronously, getting the globals symbol table, installing packages, loading files, etc. which each runtime is responsible for implementing on its own.

- For the file -- `pyconfig.ts`, `pyodide.ts` and `runtime.ts`

Usage of `initializers`, `postInitializers`, `scriptsQueue`, etc. is moved to `runtime.ts` along with the `PyodideRuntime` class which now extends from the `Runtime` parent class. This is because all code pertaining to the pyodide runtime is in one place now, not polluting / mixing up with the functionality of `PyConfig`.

### CODE REVIEW

**Comment by ntoll** [comment]
I've just taken a first glimpse. Initial thought... can we possibly add comments..? :wink:

File: pyscriptjs/src/components/base.ts
Comment (fpliger) [inline pyscriptjs/src/components/base.ts:12]: This is a good first step towards the direction we want to go (a generic runtime). 

On thing that will need to follow (in a follow up PR, not now! let's get it working first) is replacing this with a data structure that can allow use to support multiple runtimes.

File: pyscriptjs/src/components/base.ts
Comment (fpliger) [inline pyscriptjs/src/components/base.ts:133]: I'm not sure we should differentiate `runAsync` from `run` and least not for the interface or a runtime API and not how it's intended here.

My understanding is that from the `pyodide` point of view we don't need to (@rth @hoodmane, please keep me honest here) and from a Facade API point of view (which is the exercise here) I'd rather have one `run` method that accepts a generic `options` mapping that can add flexibility to add specifics to each runtime so that underneath you can decide whether to run pyodide `runPython` or `runPythonAsync`. 

That said, I'd still add a `runAsync` in the future but that would probably mean that we are running the runtime on a webworker. So basically `run` runs on the main thread while `runAsync` can be a higher level interface that we use to run things on webworkers.

/cc @ntoll @antocuni  for your 2 cents :)

File: pyscriptjs/src/components/base.ts
Comment (fpliger) [inline pyscriptjs/src/components/base.ts:176]: Having the attribute feels more Pythonic. What's the reason you decided to go with an explicit getter/setter method?

File: pyscriptjs/src/components/pyconfig.ts
Comment (fpliger) [inline pyscriptjs/src/components/pyconfig.ts:8]: This also seems more like a PR comment rather than a code comment. 

Here you should describe in a more generic way "what's going on" so that people reading the code doesn't have to guess (without context)

File: pyscriptjs/src/components/pyenv.ts
Comment (fpliger) [inline pyscriptjs/src/components/pyenv.ts:3]: same comment here...

File: pyscriptjs/src/components/pyscript.ts
Comment (fpliger) [inline pyscriptjs/src/components/pyscript.ts:10]: same comment  here

File: pyscriptjs/src/components/pyscript.ts
Comment (fpliger) [inline pyscriptjs/src/components/pyscript.ts:25]: I think we should call it runtime. It's not a promise anymore.

File: pyscriptjs/src/pyodide.ts
Comment (fpliger) [inline pyscriptjs/src/pyodide.ts:137]: Seems like we could move some of these to Runtime.initialize but we can do this in a follow up PR. Just wanting to comment something that we can improve later

File: pyscriptjs/src/runtime.ts
Comment (fpliger) [inline pyscriptjs/src/runtime.ts:11]: It's ok as first step.... we should change this soon to reflect that pyodide is the default rather the only one

File: pyscriptjs/src/runtime.ts
Comment (fpliger) [inline pyscriptjs/src/runtime.ts:26]: Right now it's an interface... I think there's some logic that we can provide to all the runtimes that subclass from it though

File: pyscriptjs/src/stores.ts
Comment (fpliger) [inline pyscriptjs/src/stores.ts:15]: This should be a generic comment not a reference to pyodide because it makes it harder for a dev without context to understand

**Review by fpliger — CHANGES_REQUESTED** [review]
Thanks @madhur-tandon ! This is a good first step in the right direction. 

The main thing I think that it's missing is tests :) (not that we have the infra and boilerplate in place we should enforce tests on every PR) and them some of the comments (but these should be pretty straightforward)

File: pyscriptjs/src/components/base.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/base.ts:12]: Yeah, we can do this later, thanks!

File: pyscriptjs/src/components/base.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/base.ts:176]: Hmm, all this is JS code, but yeah, attribute feels better, changing, thanks!

File: pyscriptjs/src/components/pyconfig.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/pyconfig.ts:8]: done, thanks!

File: pyscriptjs/src/components/pyenv.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/pyenv.ts:3]: removed!

File: pyscriptjs/src/components/pyscript.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/pyscript.ts:10]: removed!

File: pyscriptjs/src/components/pyscript.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/pyscript.ts:25]: nice nitpick 🗡️

File: pyscriptjs/src/pyodide.ts
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:137]: Yeah, I also thought the same, but wanted to have a first pass on the PR first.

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts:11]: done, removing a line about `design is based on Pyodide`

File: pyscriptjs/src/stores.ts
Comment (madhur-tandon) [reply pyscriptjs/src/stores.ts:15]: removed the reference to `pyodide`

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts:26]: yes, we can move all `initializers`, `postInitializers`, `scriptsQueue`, etc. here, done that

File: pyscriptjs/src/pyodide.ts
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:137]: moved `initializers`, `postInitializers`, `scriptsQueue`, etc. to `initialize` in super class 😄

File: pyscriptjs/src/components/base.ts
Comment (ntoll) [reply pyscriptjs/src/components/base.ts:128]: This feels like we only needs a `run` method, with perhaps (at a later date) a flag like `webworker=true` to make it run "async". Like @fpliger I prefer a single Zen-o-Python one-and-obvious way to do this, and I think a future flag should be explicit about what it's providing (i.e. we call it `webworker` rather than `async` because the former is more descriptive, even though the latter is "true").

File: pyscriptjs/src/runtime.ts
Comment (ntoll) [inline pyscriptjs/src/runtime.ts:98]: Why not just `run` ..? That the argument is called `code` tells us what it's doing. :wink:

File: pyscriptjs/src/runtime.ts
Comment (ntoll) [inline pyscriptjs/src/runtime.ts:104]: Remove this in light of the discussion elsewhere in this PR..? (We can always add it later).

My concern is that we retain the simplest possible code base and NOT add stubs that we're likely not going to use, only to forget about them later.

File: pyscriptjs/src/runtime.ts
Comment (ntoll) [inline pyscriptjs/src/runtime.ts:131]: What does the argument name `s` stand for. I was expecting the argument to be called something like `path`. Thoughts..?

File: pyscriptjs/src/runtime.ts
Comment (ntoll) [inline pyscriptjs/src/runtime.ts:142]: What does `a` stand for..? It's not clear (and feels a bit arbitrary).

**Review by ntoll — COMMENTED** [review]
This is a great first step. I think we're going to learn a LOT when we try it with a different Python engine (MicroPython!?). :tada: 

My comments annotated to the code are more about clarifications and making things explicit and easy to understand so someone unfamiliar with the codebase easily understands what's going on. :+1:

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts:98]: Yeah, thanks!

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts:131]: Thanks, it is carried forward from the existing codebase, but we should change it yes

**Comment by hoodmane** [comment]
> This feels like we only needs a run method, with perhaps (at a later date) a flag like webworker=true to make it run "async"

Personally I don't think the same method should sometimes return a Promise and sometimes return a value *particularly* in the case that the value might also be a promise. For instance the Python built in `exec` does this: if the code includes top level await it automatically wraps the result in Future. If the code doesn't contain top level await, the result might be a Future anyways. If for instance you are writing a repl, in the first case you want to await the Future and return the result. In the second case the Future is the result. It's quite annoying to distinguish between these two.

I think it's better to have different sync and async functions and combining them didn't

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts:142]: This is also carried forward from an existing codebase, do you have any ideas here? I am thinking of `default` maybe?

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts]: Still waiting on some consensus before we decide to remove this / merge this and non-async version.

**Comment by ntoll** [comment]
@hoodmane totally get what you're saying. But we're exploring ways in which to use a message bus for communication between main thread and webworker[s]. In which case, no promise (presumably to be resolved when the computation provides a result) need be returned. Rather, just listen for the expected messages on the queue. The webworker flagged `run` call becomes a file and forget.

Nota bene: this is VERY early days, and such thinking might (i.e. probably) change, in which case considerations such as the one you bring up are very much in play (and FWIW I agree with you). TL;DR - we're experimenting and exploring concepts to refine things to a consistent, cleanly abstracted and relatively simple mode of working with different Python "engines".

Does that make sense..?

File: pyscriptjs/src/runtime.ts
Comment (ntoll) [reply pyscriptjs/src/runtime.ts:142]: I'm not sure what it could be... although I imagine @fpliger probably knows.

I originally wondered if `id` could be something with an id defined in the DOM (e.g. `<div id="foo">...</div>`). Or it could just be some unique ID used internally by pyscipt... in which case something like a UUID or auto-incremented int might be better than a hard-coded 'a' for all case. :stuck_out_tongue:

File: pyscriptjs/src/components/base.ts
Comment (fpliger) [reply pyscriptjs/src/components/base.ts]: The reason I was thinking about using `run` and `runAsync` (or maybe `runOnWorker` or something like that) is because of what you and @hoodmane mentioned below.... The 2 have **significant** differences and I was expecting for it to be too much to cramp everything in 1 holy method. 

That said, I think we should defer the decision until we actually have a plan on how we want to support messaging between the worker and the main thread, how we'll support writing to the dom, etc...

File: pyscriptjs/src/runtime.ts
Comment (fpliger) [reply pyscriptjs/src/runtime.ts]: +1 on @ntoll suggestion

File: pyscriptjs/src/runtime.ts
Comment (fpliger) [reply pyscriptjs/src/runtime.ts:142]: `a` stands for `a barf and it should be "default"` 😛 

There's a lot that we need to change on that API, the mix and match of [bad] names is one of them... In this case `id` is the ID of a specific runtime instance (that I was calling "environment"). Imho, this `id` should be `default` by default and be something else specific by the user when specifying their runtime in the app config

**Comment by fpliger** [comment]
Overall, I agree with both @hoodmane and @ntoll . We should not have a function that returns a value or a `promise` depending on the weather... but also we may be able to avoid that considering the machinery to make the execution on the webworker and the main thread talk. 

Overall I think this is a first step and we need to do research and explore what others are doing in the JS world before we can design this part of the API

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts:142]: changed to `default`, thanks!

File: pyscriptjs/Makefile
Comment (fpliger) [inline pyscriptjs/Makefile:43]: Can you put a comment here saying what is it doing and why we need this?

File: pyscriptjs/__mocks__/fileMock.js
Comment (fpliger) [inline pyscriptjs/__mocks__/fileMock.js:1]: Same here.. why do we need this?

File: pyscriptjs/package.json
Comment (fpliger) [inline pyscriptjs/package.json:15]: same here... (it's just good to socialize why we **need** to run with cros-env)

File: pyscriptjs/src/components/base.ts
Comment (fpliger) [inline pyscriptjs/src/components/base.ts:199]: I don't think we need this if here anymore... do we?

Comment (fpliger) [reply pyscriptjs/src/components/base.ts:199]: Oh, nevermind... we do (for now). This is very specific to pyodide so it'd be good to move this custom logic to check if the runtime is "ready" to the runtime abstraction itself. Like adding a `isReady` method that each runtime needs to implement.

File: pyscriptjs/src/components/base.ts
Comment (fpliger) [inline pyscriptjs/src/components/base.ts:243]: same comment here

File: pyscriptjs/src/components/base.ts
Comment (fpliger) [inline pyscriptjs/src/components/base.ts:258]: This is going to be a problem in the future when we add different runtimes. Nothing to do in this PR but just for us to be aware.

File: pyscriptjs/src/components/base.ts
Comment (fpliger) [inline pyscriptjs/src/components/base.ts:264]: This is going to be a problem in the future when we add different runtimes. Nothing to do in this PR but just for us to be aware.

File: pyscriptjs/src/pyodide.ts
Comment (fpliger) [inline pyscriptjs/src/pyodide.ts:27]: Would be great to allow configuration of this and other things through the PyScript App/Runtime configuration  itself. (Can be in a follow up PR, but we should open an issue for this)

File: pyscriptjs/src/runtime.ts
Comment (fpliger) [inline pyscriptjs/src/runtime.ts:15]: I'd add that it's referring to the current **pyodide** release and not PyScript release

File: pyscriptjs/src/pyodide.ts
Comment (fpliger) [inline pyscriptjs/src/pyodide.ts:51]: Almost feels like it's an unnecessary double layer of abstraction. If `Runtime` already allows for th injection of a runtime through `interpreter`, do we need to also subclass and reimplement these methods here?

This PR is a good first step towards building a common interface, so we don't need to figure this out in this context but we can think about this after we merge.

File: pyscriptjs/src/utils.ts
Comment (fpliger) [inline pyscriptjs/src/utils.ts:87]: docstrings :)

**Review by fpliger — APPROVED** [review]
Left some comments mostly for follow up improvements. Overal the PR LGTM. Great job @madhur-tandon   🚀   TY!

**Comment by fpliger** [comment]
In some of the comments above, would be useful link to the [upstream issue](https://github.com/pyodide/pyodide/issues/2764) in the code comments

File: pyscriptjs/Makefile
Comment (madhur-tandon) [reply pyscriptjs/Makefile:43]: done, thanks

File: pyscriptjs/__mocks__/fileMock.js
Comment (madhur-tandon) [reply pyscriptjs/__mocks__/fileMock.js:1]: done, thanks

File: pyscriptjs/src/components/base.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/base.ts:258]: noted.

File: pyscriptjs/src/components/base.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/base.ts:264]: noted.

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts]: fixed it, the newer version exports it

File: pyscriptjs/src/utils.ts
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:87]: done

File: pyscriptjs/src/pyodide.ts
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:51]: noted.

File: pyscriptjs/src/pyodide.ts
Comment (madhur-tandon) [reply pyscriptjs/src/pyodide.ts:27]: noted.

File: pyscriptjs/package.json
Comment (madhur-tandon) [reply pyscriptjs/package.json:15]: JSON files cannot have comments unfortunately

File: pyscriptjs/src/components/base.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/base.ts:199]: will do it in follow up PR if it's okay?

File: pyscriptjs/src/components/base.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/base.ts:243]: for follow up PR if it's okay

File: pyscriptjs/src/components/base.ts
Comment (madhur-tandon) [reply pyscriptjs/src/components/base.ts]: decided on `runPythonAsync` 🗡️

File: pyscriptjs/src/runtime.ts
Comment (madhur-tandon) [reply pyscriptjs/src/runtime.ts]: decided on `runPythonAsync` 🗡️

**Comment by hoodmane** [comment]
Hopefully we'll release Pyodide 0.21.2 with fixes for the Jest problems in the next day or two.

**Pull request merged**
